### ISSUE TITLE
Create Manager Workflow

### ISSUE THREAD

**Issue by shawnrice** [shawnrice/alfred-bundler#25]
We need to create a workflow that will interface with the registry.json file to provide information about installed assets and allow the user to uninstall orphaned assets.

It should work across all versions of the bundler.

It could also force update the bundler between minor versions.

Another possibility is to bake in some extra functionality to delete orphaned data and cache directories.

**Comment by deanishe** [comment]
Should this be a workflow or part of the update script?

Workflow = manual management, which is normally not the best route to a well-maintained system…

**Comment by shawnrice** [comment]
The part of automatic updates in the background should be and is (untested) part of the bundler.

One problem with having the bundler is that you might install a workflow that uses one asset and get rid of the workflow, but then the asset is still there. The workflow would be to easily identify the orphaned assets and 86 them. The same interface would be there to give you information as to what workflows are using the bundler, what assets they're using, and what assets are installed.

While none of this should really matter, I feel that having a convenient way to intervene manually with the bundler will foster some nicer transparency but also make people happier overall with it.

The option to force a minor update, or we could even have to reinstall the bundler, would be there to push past any bugs that might have crept in, or, more likely, to let people feel that they have more control of it. I mean, they do, they just might not feel that they do....

**Comment by deanishe** [comment]
> I feel that having a convenient way to intervene manually with the bundler will foster some nicer transparency but also make people happier overall with it.

I reckon that the bundler staying as out of their face as possible would foster the best possible feeling towards the bundler.

No, we shouldn't install stuff on their system without asking, but apart from that, the bundler should be pretty much silent. To quote Futurama:

> If you've done your job right, nobody will know you've done anything at all.

Updating should probably involve the user, but I think the bundler should take care of deleting unused assets/libraries itself. It shouldn't mess with actual workflow data (i.e. workflows' own data/cache directories), but should periodically check which workflows have been uninstalled and remove the workflow assets specific to those workflows.

**Comment by shawnrice** [comment]
Two approaches:
1. Delete anything orphaned
2. Leave everything so that future workflows don't need to download it.

The second needs manual intervention to happen in order to delete anything. We could make it easier in that, whenever the `update` script is run (once per week), it also runs a sort of `asset_report` script that
1. Cleans the registry to remove references to deleted workflows
2. Maintains an easy to read text file of orphaned assets

If we take the first approach, then the `asset_report` script would instead delete the orphaned resources and log their deletion rather than writing a report. Right now, icons aren't registered, so they will always be left over.

If we take any sort of automated approach, then we should remove any workflow specific files/directories. Python creates these, so those should be gone. Using composer with PHP does the same, so those should be gone, although composer assets aren't registered individually in the registry yet.

Whatever approach we take, the `cleanup` script needs to be written in a language that handles JSON. Also, it needs to be forked so as not to interfere with performance.

**Comment by deanishe** [comment]
The cleanup script should be written in Python, imo. I have code for reading the bundle IDs of all installed workflows and for forking lying around already, and it's a small step to delete the workflow-specific assets/cache directories of removed workflows from there.

OTOH, I don't know anything about the registry format. If that's complicated to deal with, it might be smarter to use PHP (for which, I presume, that code already exists).

What I would recommend algorithm-wise:
1. Get list of installed workflow bundle IDs
2. Go through the `{ab_data_dir}/workflows/<bundleid>` directories, deleting the folders (containing Python libs, Ruby gems, PHP libs etc.) that belong to workflows that are no longer installed.
3. Go through all installed common assets that are not registered for installed workflows and delete the ones whose last access times are more than a month (two months?) ago.

The script would log its output to the standard bundler log file.

**Comment by shawnrice** [comment]
I've written that code a couple of times for PHP too. But go for it in Python. But, to fork it, just call the script from `update-wrapper.sh` with the same syntax that the other one is called with. That's the easiest and most reliable forking mechanism that works from 10.6 to 10.10. Then, you don't have to worry about forking the python.

Just load up `registry.json` and take a look at the contents. It's pretty damn simple. Basically:

```
asset-name
  - version
    - list of bundleids that use it
```

So, you'd just have to cycle through that and compare to the currently installed workflows. If you deregistered all the workflows for a version, then delete the version. If you delete all the versions, delete the asset. Done.

For the composer functions, just look in `{ab_data_dir}/data/assets/php/composer/bundles/{bundleids}` for the workflow directories to delete. Also, look in `{ab_cache_dir}/php/composer/{bundleid}.json`. Individual composer assets aren't registered yet, but those directories exist.

Ruby gems currently aren't registered.

Lets go with two months.

**Comment by deanishe** [comment]
> That's the easiest and most reliable forking mechanism that works from 10.6 to 10.10. Then, you don't have to worry about forking the python.

Err, no. When I say I have the technology to fork in Python, I _really_ mean it. Not this `nohup` shit. Proper [change-to-the-appropriate-working-directory-and-start-a-proper-daemon-process stuff](https://github.com/deanishe/alfred-workflow/blob/master/workflow/background.py#L144). 

> Lets go with two months.

It's a reasonable time to wait.

**Comment by shawnrice** [comment]
Okay. PHP is supposed to have [forking technology built into it](http://php.net/manual/en/function.pcntl-fork.php), but Apple decided to exclude that set of functions at compile time. I had somewhat assumed that they had done something similar with others. We need to see if they took anything out of Ruby, either 1.8 or 2.0.

**Comment by deanishe** [comment]
Nah. Python/Ruby are seen as basically system languages, not web ones, so different rules apply.
